### DISCUSSION TITLE
npm, npx are not in $PATH of Formula build environment

### DISCUSSION THREAD

**Discussion by stevehartwell** [Homebrew/discussions#1699] [Writing Formulae/Casks] [answered]
I have a Formula that `depends_on "node"` and runs `system "make", "build"` in the install section.  The build fails because the commands "npm" (and possibly "npx") in the Makefile are not found in the PATH of the build environment that `brew install` constructs for formulas.

Running `brew install -s --HEAD --interactive my-formula`, I see this in PATH:
`my-formula-20210617-21218-zwgnvy % echo $PATH`
`/Volumes/NoSave/Homebrew/Library/Homebrew/shims/mac/super:/Volumes/NoSave/Homebrew/opt/go/bin:/Volumes/NoSave/Homebrew/opt/node/bin:/usr/bin:/bin:/usr/sbin:/sbin`
`my-formula-20210617-21218-zwgnvy % node -v`
`v16.3.0`
`my-formula-20210617-21218-zwgnvy % which npm npx`
`npm not found`
`npx not found`

The `.../opt/node/bin` in PATH looks encouraging but the "npm" and "npx" commands are not there, only "node".

I tried inserting `ENV.append_path "PATH", "#{Formula["node"].libexec}/bin"` before the `system "make", "build"` line but that had no effect on the PATH of the build environment.

What can I put in my Formula that will get this Makefile to work without having to patch it?

**Comment by carlocab** [comment]
Other formulae that use `npm` do
```ruby
require "language/node"
```
at the start of the formula file. Do you have this?

**Reply by stevehartwell** [reply]
I tried it, but it didn't fix the issue.

Thanks for your suggestion,

/Steve

**Comment by stevehartwell** [comment, marked as answer]
> I tried inserting `ENV.append_path "PATH", "#{Formula["node"].libexec}/bin"` before the `system "make", "build"` line but that had no effect on the PATH of the build environment.

Actually, this _did_ fix the first problem, it just hadn't been applied yet at the `--interactive` stage that I was inspecting.  Running `brew install -d -v ...` verified my first fix.

The _second_ problem was that the upstream `Makefile` doesn't support `make`'s parallelism (which is enabled by default), and the build happened to trip over some parallelized `npm` command that made me think that my `PATH` fix wasn't working.

Adding `ENV.deparallelize` fixed the second, and hopefully final issue.
